torchtext/experimental/datasets/text_classification.py

```diff
@@ -1,93 +1,479 @@
-    if not tokenizer:
-        tokenizer = get_tokenizer("basic_english")
 
-    if not set(data_select).issubset(set(('train', 'test'))):
-        raise TypeError('Given data selection {} is not supported!'.format(data_select))
+def build_vocab(data, transforms):
+    tok_list = []
+    for (label, txt) in data:
+        tok_list.append(transforms(txt))
+    return build_vocab_from_iterator(tok_list)
+
+
+def squential_transforms(*transforms):
+    def _forward(txt_input):
+        for transform in transforms:
+            txt_input = transform(txt_input)
+        return txt_input
+    return _forward
+
+
+class TextClassificationDataset(torch.utils.data.Dataset):
+    """Defines an abstract text classification datasets.
+       Currently, we only support the following datasets:
+             - AG_NEWS
+             - SogouNews
+             - DBpedia
+             - YelpReviewPolarity
+             - YelpReviewFull
+             - YahooAnswers
+             - AmazonReviewPolarity
+             - AmazonReviewFull
+    """
 
-    dataset_tar = download_from_url(URLS[dataset_name], root=root)
-    extracted_files = extract_archive(dataset_tar)
+    def __init__(self, data, vocab, transforms):
+        """Initiate text-classification dataset.
 
-    iters_group = {}
-    if 'train' in data_select:
-        iters_group['vocab'] = _imdb_iterator('train', extracted_files,
-                                              tokenizer, ngrams)
-    for item in data_select:
-        iters_group[item] = _imdb_iterator(item, extracted_files,
-                                           tokenizer, ngrams, yield_cls=True)
-    return iters_group
+        Arguments:
+            data: a list of label and text tring tuple. label is an integer.
+                [(label1, text1), (label2, text2), (label2, text3)]
+            vocab: Vocabulary object used for dataset.
+            transforms: a tuple of label and text string transforms.
+        """
+
+        super(TextClassificationDataset, self).__init__()
+        self.data = data
+        self.vocab = vocab
+        self.transforms = transforms  # (label_transforms, tokens_transforms)
+
+    def __getitem__(self, i):
+        label = self.data[i][0]
+        txt = self.data[i][1]
+        return (self.transforms[0](label), self.transforms[1](txt))
+
+    def __len__(self):
+        return len(self.data)
+
+    def get_labels(self):
+        labels = []
+        for item in self.data:
+            label = item[0]
+            labels.apppend(self.transforms[0](label))
+        return set(labels)
+
+    def get_vocab(self):
+        return self.vocab
 
 
 def _setup_datasets(dataset_name, root='.data', ngrams=1, vocab=None,
-                    removed_tokens=[], tokenizer=None,
-                    data_select=('train', 'test')):
+                    tokenizer=None, data_select=('train', 'test')):
+    text_transform = []
+    if not tokenizer:
```

Comment:
"if tokenizer is None" is more precise. This will also succeed for empty lists and bool values, which isn't the default.